File: easy/problems/0367-valid-perfect-square/solution.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        if (num < 1) return false;

        long left = 1, right = num;
        while (left <= right) {
            long mid = (left + right) / 2;
            long square = mid * mid;

            if (square == num) {
                return true;
            } else if (square < num) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return false;
    }
};

File: easy/problems/0367-valid-perfect-square/solution.py
def isPerfectSquare(num):
    """
    Determine if a number is a perfect square.

    Args:
        num: int - the number to check

    Returns:
        bool - true if num is a perfect square
    """
    if num < 1:
        return False

    left, right = 1, num
    while left <= right:
        mid = (left + right) // 2
        square = mid * mid

        if square == num:
            return True
        elif square < num:
            left = mid + 1
        else:
            right = mid - 1

    return False
